### PR TITLE
Add ability to compute density of states (Lanczos), simpson integrator and new spslicer

### PR DESCRIPTION
**Pull request by AerysL**
This pull adds the ability to compute density of states of a matrix by the Lanczos algorithm (landos.c), a function to compute the integrals using simpson's rule and the trapezoidal rule, and a new interval partitioner.

The following files are edited or added:

- evsl.h: Include the new functions (LanDos, spslicer2, simpson, randn_double(generates a normally distributed vector)), in addition to minor spacing automatically done by my editor.
- landos.c: contains the code for the LanDos function
- makefile: add support for new functions and files
- spslice2.h: contains the code for the spslicer2 function
- vect.c: add randn_vector via Box-Muller transformation
- simpson.c: contains the code for simpson function
- LanDos.c: tests the landos, spslicer2 and simpson functions
- testMat.dat: used within the test

### CODE REVIEW

**Review by liruipeng — APPROVED** [review]
look good to me.

File: EVSL_1.0/SRC/landos.c [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+  const double kappa = 1.25;
+  const int M = min(msteps, 30);
+  const double H = (lM - lm) / (M - 1);
+  const double sigma = H / sqrt(8 * log(kappa));
+  double sigma2 = 2 * sigma * sigma;
+  // If gaussian small than tol ignore point.
+  const double tol = 1e-08;
+  double width = sigma * sqrt(-2 * log(tol));
+  linspace(aa, bb, npts, xdos);       // xdos = linspace(lm,lM, npts);
+  memset(y, 0, npts * sizeof(y[0]));  // y = zeros(size(xdos));
```
Comment (liruipeng) [inline EVSL_1.0/SRC/landos.c]: Calloc(y,...). So no need for memset(y,0,...)

Comment (AerysL) [reply EVSL_1.0/SRC/landos.c]: Fixed, thanks

File: EVSL_1.0/SRC/landos.c [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+    t = 1.0 / sqrt(t);  // t is now the number required to normalize vector.
+    DSCAL(&n, &t, v, &one);
+    DCOPY(&n, v, &one, V,
+          &one);  // v = w/norm(w); Might be able to use DNRM2 instead.
+    double wn = 0.0;
+    /*-------------------- main Lanczos loop */
+    int j;
+    for (j = 0; j < msteps; j++) {
+      // w = A*v
+      matvec_genev(A, &V[j * n], &V[(j + 1) * n]);
```
Comment (AerysL) [inline EVSL_1.0/SRC/landos.c]: matvec_genev was removed in the EVS_Bsol, which I have not yet taken into account. I will have to update my code to accommodate this change.

**Review by liruipeng — CHANGES_REQUESTED** [review]
I take my approval back... it seems that LanDOS only works for standard eig val problem, or B is factored, so you work on L^{-1}*A*L^{-T}. This might be why you call matvec_genev. Have you considered generalized eig problem without Cholesky factorization (with only B-sol)? It will need to change normal inner product to B-inner product. But anyway, the current code is in good shape and should be easy to accommodate B-Sol.

--------------
I might be wrong in my above comment. If these are only for standard eig problem, they are fine. You just to need to change  matvec_genev to matvec_A, which computes y=A*x. Yousef has matlab code for generalized problem as well. So, I thought this one is for everything.

**Comment by AerysL** [comment]
The Matlab script provided to me only supported the standard eig problem. As such I've just updated the code to support the evsldata format, and improved the tests; thanks for pointing out the changes required. I believe my next task will likely be to generalize it, so your comments will help me complete that.

**Pull request merged**